### DISCUSSION TITLE
Create decoder for Apollo Ultrasonic Smart liquid monitor

### DISCUSSION THREAD

**Discussion by StarMonkey1** [merbanan/rtl_433#2244] [Ideas]
Hi, hope someone can offer some help/advice.

RTL_433 seems to support older models of ultrasonic liquid monitors (protocol 43, 81 and 83) but not "newer" ones (actually still 6 or so years old). Specifically (https://dunravensystems.com/apollo-smart-heating-oil-energy-monitor/) which is a monitor that's rebranded and sold world wide.

I can "see" (and hear) the data through SDR# and SDR Console but RTL_433 doesn't appear able to decode it no matter which protocol I use.

I know very little about SDRs/RF so don't really know where to start in trying to build a decoder (or even if it's possible, the decoder for protocol 43 mentions Apollo SMART as not supported but not why, https://github.com/merbanan/rtl_433/blob/master/src/devices/oil_standard.c).

Does anyone have any insight into this protocol or whether anyone could offer some support in building a decoder, I can provide any data required. My end goal is to be able to monitor liquid levels within Home Assistant (just like here: https://community.home-assistant.io/t/oil-tank-level-monitor-watchman-sonic-rtl-433-integration-for-rf-sensors-or-level-sensors/223986).

This is what I see when I put the monitor into debug mode (so it regularly sends data for a set period).

![sdr](https://user-images.githubusercontent.com/104598361/203378494-37e45fec-17f3-4e1c-97ef-e982f1506de7.png)

Thanks

**Comment by zuckschwerdt** [comment]
There is currently one more level monitor in development: "Watchman Plus" #2133 
Perhaps that's already the protocol you'd need. Either way, capture some samples and upload here as zip.
https://triq.org/rtl_433/ANALYZE.html

**Comment by StarMonkey1** [comment]
Thanks for the quick reply. They're certainly related in terms of product but I've no idea if they are using the same protocol.

I'll capture some samples and get them uploaded!

**Comment by StarMonkey1** [comment]
Ok, I've grabbed a sample. It seems to show data on triq.org. I should be around more often now to supply any info as well.

[file_433.92M_250k.zip](https://github.com/merbanan/rtl_433/files/10186286/file_433.92M_250k.zip)

**Comment by zuckschwerdt** [comment]
That signal looks really good!

There is FSK PCM data with 500 µs symbol width, the payload is MC (500 µs half-bit width)

```
rtl_433 -R 0 -X 'n=name,m=FSK_PCM,s=500,l=500,g=1500,r=9000' file_433.92M_250k.cu8
```

Gives you the raw data of 143x packets (every half second) as
`{24}555558, {130}9999996a6559aaa99995a65aa665596a0, {4}f`

That's a preamble with a desync, then MC data and a postamble.

You can also accept the longer spaces (gaps) in pre-/postamble to get the code in one line, and at the same time align at the preamble:
```
rtl_433 -R 0 -X 'n=name,m=FSK_PCM,s=500,l=500,g=2000,r=9000,preamble=5558' file_433.92M_250k.cu8
```

Use the resulting code of `{130}9999996a6559aaa99995a65aa665596a0` in [a BitBench like this](https://triq.net/bitbench#c=%7B130%7D9999996a6559aaa99995a65aa665596a0%20%5BFirst%20test%5D&f=8h8h%20&d=MC&cw=4), collect more, annotate, then we can figure out what the field values are.

To collect run e.g.
```
rtl_433 -R 0 -X 'n=name,m=FSK_PCM,s=500,l=500,g=2000,r=9000,preamble=5558'
```

test various distances or temperature and such with the sensor and post the codes with your notes.

**Reply by StarMonkey1** [reply]
Ok, so it looks like I completed step 1 ok!

I'm completely new to RTL so I might need instructions step by step. I can run 
rtl_433 -R 0 -X 'n=name,m=FSK_PCM,s=500,l=500,g=2000,r=9000,preamble=5558' file_433.92M_250k.cu8 and post the results?

However if you wanted to know what the "real" data is to match up with the RF signal then I've got no way to do that. The sensor is a liquid level sensor (installed within an oil tank). I believe it will report distance to liquid surface in cm or similar, however the receiver unit you get with the measure/sender doesn't give you the raw data. It gives you total volume of liquid in the tank in litres. The raw data isn't available.

**Reply by zuckschwerdt** [reply]
Yes, I meant real data with notes to what that data might indicate (how many cm was the data you posted above?)
Decoding what you posted we get `aaa7 42fe a8d3 d427` -- we do know that is correct, high quality data (MC coding) but we can't really spot any meaning there?

**Reply by zuckschwerdt** [reply]
Or more practical: Is the unit inside the tank or screwed ontop? I my case I just unscrewed it, held it at 200cm to a wall, noted the resulting codes, then at 199cm, 198, … 190, 180 , 150, 100. The resulting codes then already revealed the mapping.

If that is not possible maybe wait until the readout goes down one count (what is that 1 L?) and capture a code again, repeat maybe 5 times and post codes with the readouts.

**Reply by StarMonkey1** [reply]
I think the only way to do it is to remove the sensor from the tank and as you say measure a distance to the wall/floor. The next readout is essentially random, I don't know when it will happen (once/twice a day) and I won't know what the oil usage has been in the time between the two readings. These are all the problems I'm trying to remove by getting the raw data into Home Assistant so I can do all my own calculations!

I'll attempt to remove the sensor.

**Comment by StarMonkey1** [comment]
Ok, @zuckschwerdt I've managed to get the sensor out and take some readings. I've no idea if the sensor needed stabilisation time between readings or whether I had the sensor measured out in distance exactly right (the sensor is recessed into a housing so I had to estimate) but there seems to be data. The err (error) file is when I pointed the sensor into free space so it couldn't get a distance reading. 0cm is where I put the sensor directly on the floor (so because of the recessed housing may actually be more like 1-2cm).
[](url)
[12cm.zip](https://github.com/merbanan/rtl_433/files/10251624/12cm.zip)
[13cm.zip](https://github.com/merbanan/rtl_433/files/10251625/13cm.zip)
[14cm.zip](https://github.com/merbanan/rtl_433/files/10251626/14cm.zip)
[15cm.zip](https://github.com/merbanan/rtl_433/files/10251627/15cm.zip)
[20cm.zip](https://github.com/merbanan/rtl_433/files/10251628/20cm.zip)
[30cm.zip](https://github.com/merbanan/rtl_433/files/10251629/30cm.zip)
[40cm.zip](https://github.com/merbanan/rtl_433/files/10251630/40cm.zip)
[50cm.zip](https://github.com/merbanan/rtl_433/files/10251631/50cm.zip)
[100cm.zip](https://github.com/merbanan/rtl_433/files/10251632/100cm.zip)
[err.zip](https://github.com/merbanan/rtl_433/files/10251633/err.zip)
[0.zip](https://github.com/merbanan/rtl_433/files/10251634/0.zip)
[10cm.zip](https://github.com/merbanan/rtl_433/files/10251635/10cm.zip)
[11cm.zip](https://github.com/merbanan/rtl_433/files/10251636/11cm.zip)

**Comment by zuckschwerdt** [comment]
Very good! Run a decode over all samples:
`rtl_433 -K FILE -R 0 -X 'n=name,m=FSK_PCM,s=500,l=500,g=2000,r=9000,preamble=5558' -F json:out.json *.cu8`
and collect to [BitBench like this](https://triq.net/bitbench#c=%7B130%7D9999996a6559aaa99995a65aa665596a0%20%5BFirst%20test%5D&c=%7B134%7D9999996a6559aaa999969aa6aa9a6995fc%5B0%5D&c=%7B134%7D9999996a6559aaa99996a66aa9a5a59a3c%5B10cm%5D&c=%7B134%7D9999996a6559aaa99996a666a9a56699fc%5B10cm%5D&c=%7B134%7D9999996a6559aaa99996a65aaa66aa563c%5B11cm%5D&c=%7B130%7D9999996a6559aaa99996a65aaa66aa560%5B11cm%5D&c=%7B134%7D9999996a6559aaa99996a65aaa6599aa3c%5B12cm%5D&c=%7B134%7D9999996a6559aaa99996a65aaa5a59a5fc%5B12cm%5D&c=%7B134%7D9999996a6559aaa99996a656aa655aa9fc%5B12cm%5D&c=%7B134%7D9999996a6559aaa99996a656aa59a95a3c%5B13cm%5D&c=%7B134%7D9999996a6559aaa99996a656aa5655563c%5B14cm%5D&c=%7B134%7D9999996a6559aaa99996a5aaaa565a563c%5B14cm%5D&c=%7B134%7D9999996a6559aaa99996a5a6aa55aaa9fc%5B15cm%5D&c=%7B134%7D9999996a6559aaa99996a5a6a9a5a555fc%5B20cm%5D&c=%7B134%7D9999996a6559aaa99996a59aa9a56999fc%5B20cm%5D&c=%7B134%7D9999996a6559aaa99996a596a959a965fc%5B30cm%5D&c=%7B134%7D9999996a6559aaa99996a59aa9596a663c%5B30cm%5D&c=%7B134%7D9999996a6559aaa99996a566a69569963c%5B40cm%5D&c=%7B134%7D9999996a6559aaa99996a566a5aa6a6a3c%5B50cm%5D&c=%7B134%7D9999996a6559aaa99996a55aa5aaa6a63c%5B50cm%5D&c=%7B134%7D9999996a6559aaa99996a55aa5a9955a3c%5B50cm%5D&c=%7B134%7D9999996a6559aaa99996a55696a9a5963c%5B100cm%5D&c=%7B134%7D9999996a6559aaa99996a55696a6599a3c%5B100cm%5D&c=%7B134%7D9999996a6559aaa99996a556aaaaaa59fc%5Berr%5D&c=%7B134%7D9999996a6559aaa99996a55669965aa9fc%5Berr%5D&c=%7B134%7D9999996a6559aaa99996a5566999a6a5fc%5Berr%5D&c=%7B134%7D9999996a6559aaa99996a556665aaaa9fc%5Berr%5D&c=%7B134%7D9999996a6559aaa999969aaaaaaa9695fc%5Berr%5D&f=PRE%3F8h8h%20ID%3F8h8h%208h%20CTR%3F8d%20CM%3A8d%20CRC8h&d=MCI&cw=4).

Also `revsum` from our revdgst found a CRC-8 (0x31, 0x00) over those 8 bytes.

All looking good. The `5558` is very likely a nested preamble. Not sure how much of the `bd0156` code is type /id /or part of the counter(?). I would have guessed flags to be next but it looks like some counter. Note your first capture: the code was different there, but higher for some reason?
Then depth in neat `cm` units and finally the checksum.

I can create a decoder with this. You would then need to monitor that "counter" field for a few month and report your findings please.

**Comment by zuckschwerdt** [comment]
There is now a draft PR in #2279 -- can you check the questions and report (either here or on the PR)?

**Reply by StarMonkey1** [reply]
Amazing! I've answered what I can so far in the PR, I'm not sure what you mean by binding counter so I couldn't answer those.
